src/core/Core.js

```diff
@@ -338,6 +338,8 @@ class Uppy {
           },
           size: file.data.size || 'N/A',
           isRemote: isRemote,
+          isCheckbox: file.isCheckbox || false,
+          isFolder: file.isFolder || false,
           remote: file.remote || '',
```

Comment:
Not sure how I feel about adding `isCheckbox` to a file object. It seems it’s only used to not close the provider panel when a checkbox is selected. I think we could separate the file name and checkbox is the markup and have two different click handlers for those cases. 

Or maybe it’s better to not show checkboxes at all do multiselect on click, and then submit only when “done” is pressed?

Reply:
`isFolder` is not used at the moment, because of uppy-server support, as you mentioned? Would it make sense to do this client-side, meaning that if a folder is selected, we requests it’s content and then add all files in it? Having a file that has `isFolder: true` feels a little weird, I see that it’s supported in Google Drive, but we have to check OneDrive, Dropbox and others as well before we go this route.

Reply:
Right, `isCheckbox ` is only used to suppress `hideAllPanels` function, which is always [called](https://github.com/sadovnychyi/uppy/blob/0f12dc81877cae0e49d50975772e0695eb5fac50/src/plugins/Dashboard/index.js#L228) on `core:file-added` event. We cannot avoid calling that event, so I'm not sure about the best solution here. Maybe we shouldn't use `hideAllPanels ` at all?

> Or maybe it’s better to not show checkboxes at all do multiselect on click, and then submit only when “done” is pressed?
This would probably solve the previous problem since we won't hide the panel automatically anymore. Hopefully you could decide on this.

I use `isFolder` in my custom `checkRestrictions` function, which isn't a perfect use case to include it in this PR, but it feels odd to exclude it.

Dropbox supports isFolder as well. The idea is that server could easily process and automatically downloads thousands of files, while doing this on client side would take much more time. But it seems like the later option could work with existing uppy-server API. Maybe we could come up with a default behavior which does this on client side, but with ability to pass a custom function which will process the folder on server side instead? `onBeforeFileAdded` could possibly be used for this.

Reply:
> Would it make sense to do this client-side, meaning that if a folder is selected, we requests it’s content and then add all files in it

I am in favor of this because it requires less implementation and less added complexity. Currently when a file is sent to the server for upload, a socket channel is created and sent to the client to receive upload progress. Sending a folder for the server to download would mean various socket channels would need to be created for each file in the folder so that the client can get upload progress for each file as they happen in parallel. While this is doable, I'd rather we don't add the extra complexity if there's an alternative. 🙂 

> The idea is that server could easily process and automatically download thousands of files, while doing this on client side would take much more time.

the actual file downloads would happen on uppy-server, we would just be fetching the metadata of all the files in the selected folder on the client side. This is already what we do when you click a folder from the dropbox/google drive view on the dashboard.

Reply:
Agree, I completely forgot about sockets since I had to remove them from my server implementation (google app engine restrictions). I'll work on this. I will skip recursive folder walk to be safe, and we will need additional 'loading' state for each checkbox where it's disabled while we fetch the data.

Reply:
> Or maybe it’s better to not show checkboxes at all do multiselect on click, and then submit only when “done” is pressed?

BTW, I still like this idea but how do we differentiate between selection a folder and opening it?

Reply:
Yes, we could differentiate on double click, but that is tricky on mobile. Let’s go with checkboxes for now.